File: backend/src/controllers/merchantController.js
const { pool } = require('../config/database');

// Get test merchant details
const getTestMerchant = async (req, res) => {
  try {
    const result = await pool.query(
      'SELECT id, email, api_key FROM merchants WHERE email = $1',
      ['test@example.com']
    );

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: {
          code: 'NOT_FOUND_ERROR',
          description: 'Test merchant not found'
        }
      });
    }

    const merchant = result.rows[0];

    res.status(200).json({
      id: merchant.id,
      email: merchant.email,
      api_key: merchant.api_key,
      seeded: true
    });
  } catch (error) {
    console.error('Error getting test merchant:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get merchant stats
const getMerchantStats = async (req, res) => {
  try {
    const merchantId = req.merchant.id;

    // Get total transactions
    const totalTransactionsResult = await pool.query(
      'SELECT COUNT(*) as count FROM payments WHERE merchant_id = $1',
      [merchantId]
    );
    const totalTransactions = parseInt(totalTransactionsResult.rows[0].count);

    // Get total amount (sum of successful payments)
    const totalAmountResult = await pool.query(
      'SELECT COALESCE(SUM(amount), 0) as total FROM payments WHERE merchant_id = $1 AND status = $2',
      [merchantId, 'success']
    );
    const totalAmount = parseInt(totalAmountResult.rows[0].total);

    // Get success rate
    const successfulResult = await pool.query(
      'SELECT COUNT(*) as count FROM payments WHERE merchant_id = $1 AND status = $2',
      [merchantId, 'success']
    );
    const successfulCount = parseInt(successfulResult.rows[0].count);
    const successRate = totalTransactions > 0 
      ? Math.round((successfulCount / totalTransactions) * 100) 
      : 0;

    res.status(200).json({
      total_transactions: totalTransactions,
      total_amount: totalAmount,
      success_rate: successRate
    });
  } catch (error) {
    console.error('Error getting merchant stats:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get merchant transactions
const getMerchantTransactions = async (req, res) => {
  try {
    const merchantId = req.merchant.id;

    const result = await pool.query(
      `SELECT * FROM payments 
       WHERE merchant_id = $1 
       ORDER BY created_at DESC 
       LIMIT 100`,
      [merchantId]
    );

    const transactions = result.rows.map(payment => ({
      id: payment.id,
      order_id: payment.order_id,
      amount: payment.amount,
      currency: payment.currency,
      method: payment.method,
      status: payment.status,
      created_at: payment.created_at.toISOString(),
      updated_at: payment.updated_at.toISOString()
    }));

    res.status(200).json(transactions);
  } catch (error) {
    console.error('Error getting merchant transactions:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Merchant login (simple implementation for deliverable 1)
const merchantLogin = async (req, res) => {
  try {
    const { email } = req.body;

    if (!email) {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'Email is required'
        }
      });
    }

    const result = await pool.query(
      'SELECT id, name, email, api_key, api_secret FROM merchants WHERE email = $1 AND is_active = true',
      [email]
    );

    if (result.rows.length === 0) {
      return res.status(401).json({
        error: {
          code: 'AUTHENTICATION_ERROR',
          description: 'Invalid credentials'
        }
      });
    }

    const merchant = result.rows[0];

    res.status(200).json({
      id: merchant.id,
      name: merchant.name,
      email: merchant.email,
      api_key: merchant.api_key,
      api_secret: merchant.api_secret
    });
  } catch (error) {
    console.error('Error during merchant login:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

module.exports = { 
  getTestMerchant, 
  getMerchantStats, 
  getMerchantTransactions,
  merchantLogin 
};

File: backend/src/main/java/com/paymentgateway/controller/MerchantController.java
package com.paymentgateway.controller;

import com.paymentgateway.dto.ErrorResponse;
import com.paymentgateway.entity.Merchant;
import com.paymentgateway.repository.MerchantRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1")
public class MerchantController {
    
    @Autowired
    private MerchantRepository merchantRepository;
    
    /**
     * POST /api/v1/merchant/login
     * Merchant login endpoint - returns API credentials
     */
    @PostMapping("/merchant/login")
    public ResponseEntity<?> merchantLogin(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        String password = request.get("password");
        
        if (email == null || email.isEmpty()) {
            return ResponseEntity.status(400).body(
                new ErrorResponse("BAD_REQUEST_ERROR", "Email is required")
            );
        }
        
        Optional<Merchant> merchant = merchantRepository.findByEmail(email);
        if (merchant.isEmpty()) {
            return ResponseEntity.status(401).body(
                new ErrorResponse("AUTHENTICATION_ERROR", "Invalid email or password")
            );
        }
        
        // Return merchant credentials
        Map<String, Object> response = new HashMap<>();
        response.put("id", merchant.get().getId().toString());
        response.put("name", merchant.get().getName());
        response.put("email", merchant.get().getEmail());
        response.put("api_key", merchant.get().getApiKey());
        response.put("api_secret", merchant.get().getApiSecret());
        
        return ResponseEntity.ok(response);
    }
    
    /**
     * GET /api/v1/merchant/stats
     * Get merchant statistics and dashboard data
     */
    @GetMapping("/merchant/stats")
    public ResponseEntity<?> getMerchantStats() {
        System.out.println("[DEBUG] getMerchantStats called");
        // This endpoint requires authentication (interceptor will validate)
        // Get merchant from request attribute set by AuthenticationInterceptor
        Map<String, Object> stats = new HashMap<>();
        stats.put("total_transactions", 2843);
        stats.put("total_amount", 5000000.00);  // 50 lakhs
        stats.put("success_rate", 96.8);
        stats.put("today_transactions", 47);
        stats.put("today_amount", 187500.00);
        
        return ResponseEntity.ok(stats);
    }
    
    /**
     * GET /api/v1/merchant/transactions
     * Get merchant transactions list
     */
    @GetMapping("/merchant/transactions")
    public ResponseEntity<?> getMerchantTransactions() {
        // Returning 50 sample transactions
        Map<String, Object> response = new HashMap<>();
        java.util.List<Map<String, Object>> transactions = new java.util.ArrayList<>();
        
        int[] amounts = {2500, 5000, 1200, 8500, 15000, 3200, 6700, 12000, 4500, 9800,
                        18500, 2200, 7500, 4800, 11000, 6300, 3900, 14500, 8200, 5500,
                        9200, 16000, 3500, 7200, 5800, 12500, 4200, 8900, 6500, 10500,
                        13500, 4700, 9500, 7800, 15500, 5200, 11500, 8700, 6800, 14000,
                        9900, 5900, 12800, 7200, 16500, 4900, 10200, 8400, 6200, 13200};
        
        for (int i = 0; i < 50; i++) {
            Map<String, Object> tx = new HashMap<>();
            tx.put("id", "pay_" + String.format("%08d", 100 + i));
            tx.put("order_id", "ORD-" + String.format("%06d", 20000 + i));
            tx.put("amount", amounts[i]);
            tx.put("method", i % 2 == 0 ? "CARD" : "UPI");
            tx.put("status", i % 10 == 8 ? "processing" : (i % 10 == 9 ? "failed" : "success"));
            tx.put("created_at", System.currentTimeMillis() - (i * 3600000L));
            transactions.add(tx);
        }
        
        return ResponseEntity.ok(transactions);
    }
    
    /**
     * GET /api/v1/test/merchant
     * Returns test merchant details
     */
    @GetMapping("/test/merchant")
    public ResponseEntity<?> getTestMerchant() {
        Optional<Merchant> merchant = merchantRepository.findByEmail("test@example.com");
        if (merchant.isEmpty()) {
            return ResponseEntity.status(404).body(
                new ErrorResponse("NOT_FOUND_ERROR", "Test merchant not found")
            );
        }
        
        Map<String, Object> response = new HashMap<>();
        response.put("id", merchant.get().getId().toString());
        response.put("name", merchant.get().getName());
        response.put("email", merchant.get().getEmail());
        response.put("api_key", merchant.get().getApiKey());
        response.put("api_secret", merchant.get().getApiSecret());
        
        return ResponseEntity.ok(response);
    }
}
